### PR TITLE
Fix timestamp not passed to RecordMetadata

### LINKED ISSUE #1270: Degrade after LegacyRecords merge
As I moved the generation of a timestamp into LegacyRecordBatchBuilder the timestamp returned by future may now be None, as it was not generated yet.
This is more of a reminder for myself to fix it.

### PR DESCRIPTION
**Pull request by tvoinarovskyi**
Tried to fix #1270 and found, that it did not work properly in the first place. Was a bug in FutureRecordMetadata, that ignored timestamp=-1 being returned from the broker.

Also minor fixes include:
* Avoid getting current timestamp if magic==0 (broker <= 0.9), as it's not needed anyway. Will return -1 for this case in ```RecordMetadata```.
* Fixed timestamp error message if it's of improper type.
* Added more tests for LegacyBatch
* Added integration test for correct RecordMetadata fields, including `checksum`

### CODE REVIEW

File: test/test_producer.py [hunk 1]
```diff
@@ -78,3 +78,54 @@ def test_kafka_producer_gc_cleanup():
...
+
+@pytest.mark.skipif(not version(), reason="No KAFKA_VERSION set")
+@pytest.mark.parametrize("compression", [None, 'gzip', 'snappy', 'lz4'])
+def test_kafka_producer_proper_record_metadata(kafka_broker, compression):
+    connect_str = ':'.join([kafka_broker.host, str(kafka_broker.port)])
+    producer = KafkaProducer(bootstrap_servers=connect_str,
+                             retries=5,
+                             max_block_ms=10000,
+                             compression_type=compression,
+                             value_serializer=str.encode)
```
Comment (dpkp) [inline test/test_producer.py]: You're sending bytes below so no serializer needed here.

Comment (tvoinarovskyi) [reply test/test_producer.py]: Copy-paste)

File: kafka/record/legacy_records.py [hunk 2]
```diff
@@ -484,3 +490,37 @@ def estimate_size_in_bytes(cls, magic, compression_type, key, value):
                 cls.record_size(magic, key, value)
             )
         return cls.LOG_OVERHEAD + cls.record_size(magic, key, value)
+
+
+class LegacyRecordMetadata(object):
```
Comment (jeffwidman) [inline kafka/record/legacy_records.py]: I'm sorry, I am clearly missing something... What is the purpose of this class? It doesn't seem to do anything other than enforcing read-only on the attributes. Is there something else it does that I'm not realizing?

Comment (tvoinarovskyi) [reply kafka/record/legacy_records.py]: No, it's more or less as you said. It's just a class to store results of a function. I find this a bit more clean than `namedtuple` when it comes to optimization. This is a hot part.

It's a more convenient way of returning several values from a function. Returning a tuple will require you to parse it every time, but returning either `None` or an instance will only require a None check. See https://github.com/dpkp/kafka-python/pull/1273/files#diff-261f56705197d1ab9e6fca9249856556 and https://github.com/dpkp/kafka-python/pull/1273/files#diff-d98d7e54a13e15511c44d90e78f105e9R368.

**Pull request merged**